inc/Engine/CriticalPath/CPCSSService.php

```diff
@@ -0,0 +1,256 @@
+	public function process_generate( $item_url, $item_path, $timeout = false ) {
+		// Ajax call requested a timeout.
+		if ( $timeout ) {
+			return $this->process_timeout( $item_url );
+		}
+
+		$cpcss_job_id = $this->data_manager->get_cache_job_id( $item_url );
+
+		if ( false === $cpcss_job_id ) {
+			return $this->send_generation_request( $item_url, $item_path );
+		}
+
+		// job_id is found and we need to check status for it.
+		return $this->check_cpcss_job_status( $cpcss_job_id, $item_path, $item_url );
+	}
+
+	/**
+	 * Send Generation first request.
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $item_url Url for item to send the generation request for.
+	 * @param string $item_path Path for item to send the generation request for.
+	 * @return array
+	 */
+	private function send_generation_request( $item_url, $item_path ) {
+		// call send generation request from APIClient for the first time.
+		$generated_job = $this->api_client->send_generation_request( $item_url );
+
+		// validate generate response.
+		if ( is_wp_error( $generated_job ) ) {
+			// Failed so return back the data.
+			return $generated_job;
+		}
+
+		// Send generation request succeeded.
+		// Save job_id into cache.
+		$this->data_manager->set_cache_job_id( $item_url, $generated_job->data->id );
+
+		return $this->check_cpcss_job_status( $generated_job->data->id, $item_path, $item_url );
+	}
+
+	/**
+	 * Get job details by job_id.
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $job_id ID for the job to get details.
+	 * @param string $item_url URL for item to be used in error messages.
+	 * @return array|mixed|WP_Error
+	 */
+	private function get_cpcss_job_details( $job_id, $item_url ) {
+		$job_details = $this->api_client->get_job_details( $job_id, $item_url );
+
+		if ( is_wp_error( $job_details ) ) {
+			return $job_details;
+		}
+
+		return $job_details;
+	}
+
+	/**
+	 * Check status and process the output for a job.
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $job_id ID for the job to get details.
+	 * @param string $item_path Path for this item to be validated.
+	 * @param string $item_url URL for item to be used in error messages.
+	 * @return array|WP_Error Response in case of success, failure or pending.
+	 */
+	private function check_cpcss_job_status( $job_id, $item_path, $item_url ) {
+		$job_details = $this->api_client->get_job_details( $job_id, $item_url );
+
+		if ( is_wp_error( $job_details ) ) {
+			$this->data_manager->delete_cache_job_id( $item_url );
+			return $job_details;
+		}
+
+		if ( 200 !== $job_details->status ) {
+			// On job error.
+			return $this->on_job_error( $job_details, $item_url );
+		}
+
+		// On job status 200.
+		$job_state = $job_details->data->state;
+
+		// For pending job status.
+		if ( isset( $job_state ) && 'complete' !== $job_state ) {
+			return $this->on_job_pending( $item_url );
+		}
+
+		// For successful job status.
+		if (
+			isset( $job_state, $job_details->data->critical_path ) &&
+			'complete' === $job_state
```

Comment:
As our standard, let's build these types of long conditions like this:

```suggestion
			isset( $job_state, $job_details->data->critical_path ) 
			&&
			'complete' === $job_state
```

Why? It's easier to read. You see that's an AND and each line is a conditional.